app/domain/authentication/common/extract_resource_restrictions.rb

```diff
@@ -0,0 +1,57 @@
+require 'command_class'
+
+module Authentication
+  module Common
+
+    ExtractResourceRestrictions = CommandClass.new(
+        dependencies: {
+            role_class:     ::Role,
+            resource_class: ::Resource,
+            logger:         Rails.logger
+        },
+        inputs:   %i(authenticator_name service_id host_name account)
+    ) do
+
+      def call
+        fetch_all_role_annotations
+        convert_to_resource_restrictions
+        @resource_restrictions
+      end
+
+      private
+
+      def fetch_all_role_annotations
+        # Validate role exists, otherwise getting role annotations return empty hash.
+        raise Errors::Authentication::Security::RoleNotFound, role_id unless @resource_class[role_id]
+        role_annotations
+      end
+
+      def convert_to_resource_restrictions
+        @resource_restrictions = {}
+        @role_annotations.each do |annotation_name, annotation_value|
```

Comment:
this should be `role_annotations` instead of `@role_annotations` as it is memoized